### ISSUE TITLE
[Feature Request] Accept a list of `nn.TransformerEncoderLayer` instances as the parameter of `nn.TransformerEncoder`'s __init__ method

### ISSUE THREAD

**Issue by soloice** [pytorch/pytorch#39629]
## 🚀 Feature
Currently, an `nn.TransformerEncoder` accepts an `nn.TransformerEncoderLayer` instance in its `__init__` method and deep clones it `nlayer` times.
However, a better and more intuitive way is to pass a list of `nn.TransformerEncoderLayer` instances into an `nn.TransformerEncoder` as its internal layers.

## Motivation

Current implementation has two drawbacks:
1. If someone implements a Transformer model as follows (note the `self.` in curly brackets):
```
class TransformerModel(nn.Module):
    def __init__(self, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
        super(TransformerModel, self).__init__()
        {self.}encoder_layers = nn.TransformerEncoderLayer(ninp, nhead, nhid, dropout)
        self.transformer_encoder = nn.TransformerEncoder({self.}encoder_layers, nlayers)
```
If there's no `{self.}` (as in [official transformer tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html)), the code just works fine; but if someone adds the `{self.}`, the `TransformerModel` instance will treat `self.encoder_layers` as its submodule while not able to find gradients of its parameters during back-propagation (because `nlayers` new `TransformerEncoderLayer` instances will be deep cloned inside `nn.TransformerEncoder` and `self.encoder_layers` itself will NOT participate in the Module's forward calculation) and leads to a failure. This is exactly why I make this feature request here - I spent a lot of time on this and find the deep clone issue.

2. `nlayer` Transformer layers cannot share parameters (though one can achieve this by calling the same `nn.TransformerEncoderLayer` instance multiple times).

## Pitch

The better abstraction will be: do NOT deep clone the `nn.TransformerEncoderLayer` instance `nlayer` times. Instead, pass a length-`nlayer` `nn.TransformerEncoderLayer` instances and use them directly in the newly created `nn.TransformerEncoder`. This design resembles `RNNCell` and `MultiRNNCell` in TensorFlow.

To keep backwards compatibility, one can check the data type of `encoder_layer` in `nn.TransformerEncoder`'s `__init__` method `def __init__(self, encoder_layer, num_layers, norm=None)`. A possible implementation could be:
```
    def __init__(self, encoder_layer, num_layers, norm=None):
        super(TransformerEncoder, self).__init__()
        if isinstance(encoder_layer, list) and all([isinstance(l, TransformerEncoderLayer) for l in encoder_layer]):
            self.layers = ...   # whatever one can do
        else:
            self.layers = _get_clones(encoder_layer, num_layers)  # the original logic
```

But one has to be careful about the parameter sharing in this case (inside and outside the `nn.TransformerEncoder`), which I'm not quite familiar with.

## Alternatives
1. Also, one can provide another `__init__` method for `nn.TransformerEncoder`: pass in hyperparameters (instead of an instance) of each `nn.TransformerEncoderLayer` and construct all `nn.TransformerEncoderLayer`s directly. This can avoid the confusing behavior, too.
2. Do not produce gradient errors for unused parameters in an `nn.Module`. Print warnings instead.

**Comment by pbelevich** [comment]
@zhangguanheng66 WDYT?

**Comment by zhangguanheng66** [comment]
The transformer encoder is a pretty simple wrapper and you implement the pitch for your application. Right now, we don't want to change the API for a significant BC break.

**Comment by soloice** [comment]
@zhangguanheng66 Thanks for your reply. I suppose my pitch won't break backward compatibility.
It's also totally fine to keep it as it is, because one can easily implement what I mentioned above. Though, this thread may help people who meet the same errors (i.e.: "'NoneType' object has no attribute 'data'") as I do.
